File: backend/src/main/java/com/montanaritech/contable/facturacion/facturacompra/FacturaCompra.java
package com.montanaritech.contable.facturacion.facturacompra;

import com.montanaritech.contable.common.estado.EstadoDocumento;
import com.montanaritech.contable.common.tenant.EntidadNegocio;
import com.montanaritech.contable.contabilidad.asiento.Asiento;
import com.montanaritech.contable.contabilidad.asiento.AsientoLinea;
import com.montanaritech.contable.facturacion.TipoComprobante;
import com.montanaritech.contable.maestros.moneda.Moneda;
import com.montanaritech.contable.maestros.proveedor.Proveedor;
import com.montanaritech.contable.maestros.proyecto.Proyecto;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.ForeignKey;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OrderBy;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import lombok.Getter;
import lombok.Setter;

/**
 * Factura de compra (F4.1 §5, F4.3): simétrica a {@link
 * com.montanaritech.contable.facturacion.facturaventa.FacturaVenta}, mismo
 * modelo para notas de crédito/débito (ADR-13, signo derivado de {@code
 * tipoComprobante}). Las percepciones sufridas se guardan aparte en {@code
 * comprobante_tributo} (vínculo por {@code comprobanteTipo}/{@code
 * comprobanteId}, no una relación JPA).
 */
@Entity
@Table(name = "factura_compra", uniqueConstraints = @UniqueConstraint(
        name = "uk_factura_compra", columnNames = {"tenant_id", "proveedor_id", "tipo_comprobante", "punto_venta", "numero"}))
@Getter
@Setter
public class FacturaCompra extends EntidadNegocio {

    @ManyToOne(optional = false)
    @JoinColumn(name = "proveedor_id", nullable = false, foreignKey = @ForeignKey(name = "fk_factura_compra_proveedor"))
    private Proveedor proveedor;

    @ManyToOne(optional = true)
    @JoinColumn(name = "proyecto_id", foreignKey = @ForeignKey(name = "fk_factura_compra_proyecto"))
    private Proyecto proyecto;

    @Column(nullable = false)
    private LocalDate fecha;

    @Column(name = "fecha_vencimiento")
    private LocalDate fechaVencimiento;

    @Enumerated(EnumType.STRING)
    @Column(name = "tipo_comprobante", nullable = false, length = 20)
    private TipoComprobante tipoComprobante;

    @Column(name = "punto_venta", length = 20)
    private String puntoVenta;

    @Column(nullable = false, length = 20)
    private String numero;

    @ManyToOne(optional = false)
    @JoinColumn(name = "moneda_id", nullable = false, foreignKey = @ForeignKey(name = "fk_factura_compra_moneda"))
    private Moneda moneda;

    @Column(name = "tipo_cambio", nullable = false, precision = 19, scale = 6)
    private BigDecimal tipoCambio;

    @Enumerated(EnumType.STRING)
    @Column(name = "fuente_tc", length = 20)
    private AsientoLinea.FuenteTc fuenteTc;

    @Column(nullable = false, precision = 18, scale = 2)
    private BigDecimal neto = BigDecimal.ZERO;

    @Column(name = "importe_iva", nullable = false, precision = 18, scale = 2)
    private BigDecimal importeIva = BigDecimal.ZERO;

    @Column(name = "importe_percepciones", nullable = false, precision = 18, scale = 2)
    private BigDecimal importePercepciones = BigDecimal.ZERO;

    @Column(nullable = false, precision = 18, scale = 2)
    private BigDecimal total;

    @Column(name = "total_ars", nullable = false, precision = 18, scale = 2)
    private BigDecimal totalArs;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private EstadoDocumento estado = EstadoDocumento.BORRADOR;

    @ManyToOne(optional = true)
    @JoinColumn(name = "asiento_id", foreignKey = @ForeignKey(name = "fk_factura_compra_asiento"))
    private Asiento asiento;

    @Column(length = 2000)
    private String observaciones;

    @OneToMany(mappedBy = "facturaCompra", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("orden ASC")
    private List<FacturaCompraLinea> lineas = new ArrayList<>();
}

File: frontend/src/types/factura-compra.ts
import type { TipoComprobante } from "@/types/factura-venta"

export type EstadoFacturaCompra = "BORRADOR" | "CONFIRMADO" | "ANULADO"

export const TIPOS_TRIBUTO_COMPRA = ["PERCEPCION_IVA", "PERCEPCION_IIBB"] as const
export type TipoTributoCompra = (typeof TIPOS_TRIBUTO_COMPRA)[number]

export type FacturaCompraLinea = {
  id: number
  orden: number
  descripcion: string
  tipoCostoId: number
  tipoCostoNombre: string
  importeNeto: number
  alicuotaIva: number
  importeIva: number
  cuentaContableId: number | null
  cuentaContableCodigo: string | null
}

export type FacturaCompraTributo = {
  id: number
  tipo: TipoTributoCompra
  jurisdiccionId: number | null
  jurisdiccionNombre: string | null
  base: number | null
  alicuota: number | null
  importe: number
}

export type FacturaCompra = {
  id: number
  proveedorId: number
  proveedorNombre: string
  proyectoId: number | null
  proyectoNombre: string | null
  fecha: string
  fechaVencimiento: string | null
  tipoComprobante: TipoComprobante
  puntoVenta: string | null
  numero: string
  monedaId: number
  monedaCodigo: string
  tipoCambio: number
  neto: number
  importeIva: number
  importePercepciones: number
  total: number
  totalArs: number
  estado: EstadoFacturaCompra
  asientoId: number | null
  asientoNumero: number | null
  observaciones: string | null
  lineas: FacturaCompraLinea[]
  tributos: FacturaCompraTributo[]
}

export type FacturaCompraLineaInput = {
  descripcion: string
  tipoCostoId: number
  importeNeto: number
  alicuotaIva: number
  cuentaContableId?: number
}

export type FacturaCompraTributoInput = {
  tipo: TipoTributoCompra
  jurisdiccionId?: number
  base?: number
  alicuota?: number
  importe: number
}

export type FacturaCompraCrearInput = {
  proveedorId: number
  proyectoId?: number
  fecha: string
  fechaVencimiento?: string
  tipoComprobante: TipoComprobante
  puntoVenta?: string
  numero: string
  monedaId: number
  tipoCambio: number
  observaciones?: string
  lineas: FacturaCompraLineaInput[]
  tributos?: FacturaCompraTributoInput[]
}

export type FacturaCompraEditarInput = FacturaCompraCrearInput
